Vostok.Metrics.System/Host/DiskUsageCollector_Linux.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Vostok.Metrics.System.Helpers;
+
+namespace Vostok.Metrics.System.Host
+{
+    internal class DiskUsageCollector_Linux
+    {
+        private Dictionary<string, DiskStats> previousDisksStatsInfo = new Dictionary<string, DiskStats>();
+        private readonly Stopwatch stopwatch = new Stopwatch();
+
+        public void Collect(HostMetrics metrics, IEnumerable<string> diskstatsLines)
+        {
+            var deltaTime = (double) stopwatch.ElapsedTicks / Stopwatch.Frequency;
+
+            var disksUsageInfo = new Dictionary<string, DiskUsageInfo>();
+
+            foreach (var diskStats in ParseDiskstats(diskstatsLines))
+            {
+                var result = new DiskUsageInfo {DiskName = diskStats.DiskName};
+
+                if (previousDisksStatsInfo.ContainsKey(diskStats.DiskName))
+                    FillInfo(result, previousDisksStatsInfo[diskStats.DiskName], diskStats, deltaTime);
+
+                disksUsageInfo[result.DiskName] = result;
+
+                previousDisksStatsInfo[diskStats.DiskName] = diskStats;
+            }
+
+            metrics.DisksUsageInfo = disksUsageInfo;
+
+            RemoveUnusedEntries(disksUsageInfo);
+
+            stopwatch.Restart();
+        }
+
+        private void FillInfo(DiskUsageInfo toFill, DiskStats previousDiskStats, DiskStats diskStats, double deltaTime)
+        {
+            var readsDelta = diskStats.ReadsCount - previousDiskStats.ReadsCount;
+            var writesDelta = diskStats.WritesCount - previousDiskStats.WritesCount;
+            var timeReadDelta = diskStats.TimeSpentReading - diskStats.TimeSpentReading;
+            var timeWriteDelta = diskStats.TimeSpentWriting - diskStats.TimeSpentWriting;
+
+            if (readsDelta > 0)
+                toFill.ReadLatency = (double) timeReadDelta / readsDelta;
+            if (writesDelta > 0)
+                toFill.WriteLatency = (double) timeWriteDelta / writesDelta;
+
+            toFill.CurrentQueueLength = diskStats.CurrentQueueLength - previousDiskStats.CurrentQueueLength;
+
+            toFill.ReadsPerSecond = readsDelta / deltaTime;
+            toFill.WritesPerSecond = writesDelta / deltaTime;
+
+            toFill.IdleTimePercent = ((1 - (timeReadDelta + timeWriteDelta) / deltaTime) * 100).Clamp(0, 100);
```

Comment:
I'd like to see a proof of this method's correctness. E.g. why the sum of `timeReadDelta` and `timeWriteDelta` can't be far greater than `deltaTime` due to parallelism of I/O operations, merges, etc.

Reply:
Also, it occured to me that it would probably be cool to reverse this metric (utlilzation percent instead of idle percent) to line up with similar metric suggestion for network I/O.